### PR TITLE
Handle an unknown in a tuple correctly in reshape

### LINKED ISSUE #489: reshape doesn't match numpy semantics
```
>>> import numpy
>>> import cunumeric
>>> x = cunumeric.ones((25))
>>> x.reshape(-1, 5)
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1.])
>>> numpy.reshape(x, (-1, 5))
array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])
>>>
```

### PR DESCRIPTION
**Pull request by magnatelee**
Fixes #489

### CODE REVIEW

File: cunumeric/array.py [hunk 1]
```diff
@@ -2994,6 +2994,31 @@ def reshape(self, shape, order="C") -> ndarray:
...
+            raise ValueError("can only specify one unknown dimension")
+
+        unknown_extent = self.size
+        for extent in computed_shape:
+            if extent == -1 or extent == 0:
+                continue
+            if unknown_extent % extent != 0:
+                raise ValueError(
+                    f"cannot reshape array of size {self.size} "
+                    f"into shape ({extent})"
```
Comment (manopapad) [inline cunumeric/array.py:3010]: I would give the full shape in this error message:
```
f"into shape {computed_shape}"
```

File: cunumeric/array.py [hunk 2]
```diff
@@ -2994,6 +2994,31 @@ def reshape(self, shape, order="C") -> ndarray:
...
+        computed_shape = tuple(
+            unknown_extent if extent == -1 else extent
+            for extent in computed_shape
+        )
+
+        # Handle an easy case
+        if computed_shape == self.shape:
+            return self
+
         if shape != -1:
```
Comment (manopapad) [inline cunumeric/array.py:3022]: Most (or all) of this block is likely now redundant

File: cunumeric/array.py [hunk 3]
```diff
@@ -2994,6 +2994,31 @@ def reshape(self, shape, order="C") -> ndarray:
...
+        unknown_extent = self.size
+        for extent in computed_shape:
+            if extent == -1 or extent == 0:
+                continue
+            if unknown_extent % extent != 0:
+                raise ValueError(
+                    f"cannot reshape array of size {self.size} "
+                    f"into shape ({extent})"
+                )
+            unknown_extent //= extent
```
Comment (manopapad) [inline cunumeric/array.py:3012]: I think if you add this check here:

```
if all(extent >= 0 for extent in computed_shape) and unknown_extent != 1:
  raise ValueError(f"cannot reshape array of shape {self.shape} into shape {computed_shape}")
```

then we can get rid of the `shape != -1` block below.

File: cunumeric/array.py [hunk 4]
```diff
@@ -2994,6 +2994,31 @@ def reshape(self, shape, order="C") -> ndarray:
         --------
         Multiple GPUs, Multiple CPUs
         """
+
+        computed_shape = (shape,) if isinstance(shape, int) else shape
+
+        if sum(extent == -1 for extent in computed_shape) > 1:
```
Comment (manopapad) [inline cunumeric/array.py:3000]: NumPy will accept any negative number as an "unknown" dimension, so I would change this to
```
extent < 0 for extent in computed_shape
```

File: cunumeric/array.py [hunk 5]
```diff
@@ -2994,6 +2994,31 @@ def reshape(self, shape, order="C") -> ndarray:
...
+            raise ValueError("can only specify one unknown dimension")
+
+        unknown_extent = self.size
+        for extent in computed_shape:
+            if extent == -1 or extent == 0:
```
Comment (manopapad) [inline cunumeric/array.py:3005]: Similarly to above, I would change this to `extent <= 0`

File: cunumeric/array.py [hunk 6]
```diff
@@ -2994,6 +2994,31 @@ def reshape(self, shape, order="C") -> ndarray:
...
+                continue
+            if unknown_extent % extent != 0:
+                raise ValueError(
+                    f"cannot reshape array of size {self.size} "
+                    f"into shape ({extent})"
+                )
+            unknown_extent //= extent
+        computed_shape = tuple(
+            unknown_extent if extent == -1 else extent
```
Comment (manopapad) [inline cunumeric/array.py:3014]:
```
unknown_extent if extent < 0 else extent
```

File: cunumeric/array.py (hunk 1 above)
Comment (magnatelee) [reply cunumeric/array.py:3010]: this is to match NumPy's error message. I don't think the full shape is any more useful to identify the bug.

**Comment by manopapad** [comment]
This PR may be unnecessary. Please check the discussion on https://github.com/nv-legate/cunumeric/issues/489. The new tests should definitely be added, however, including a test using `-2` to denote an unknown dimension.

**Comment by magnatelee** [comment]
> This PR may be unnecessary. Please check the discussion on https://github.com/nv-legate/cunumeric/issues/489. The new tests should definitely be added, however, including a test using -2 to denote an unknown dimension.

Well I think this rewrite is still useful, as it made the logic cleaner and easier to understand. I'd still keep this PR even if it's functionally redundant.

**Comment by magnatelee** [comment]
Ok. I just push a fix for #127 as well.

File: cunumeric/array.py [hunk 7]
```diff
@@ -2994,61 +2993,41 @@ def reshape(self, shape, order="C") -> ndarray:
...
+        if len(args) == 0:
+            raise TypeError("reshape() takes exactly 1 argument (0 given)")
+        elif len(args) == 1:
+            shape = (args[0],) if isinstance(args[0], int) else args[0]
         else:
-            # Compute a flattened version of the shape
-            shape = (self.size,)
+            shape = args
+
+        computed_shape = tuple(operator.index(extent) for extent in shape)
```
Comment (manopapad) [inline cunumeric/array.py:3004]: Why this? Probably doesn't hurt, just curious about when this might come up.

Comment (magnatelee) [reply cunumeric/array.py:3004]: that line is exactly for the case where someone tries `arr.reshape(ext1, ext2, ext3, ...)`, which fixes #127.

File: cunumeric/array.py (hunk 3 above)
Comment (manopapad) [reply cunumeric/array.py:3018]: This is still missing. We should have a check like this to catch this case early:

```
x = cn.ones((25,))
print(x.reshape(1, 5).shape)
```

currently this fails later on:

```
Traceback (most recent call last):
  File "/Users/mpapadakis/legate.core/install/lib/python3.8/site-packages/legion_top.py", line 440, in legion_python_main
    run_path(args[start], run_name='__main__')
  File "/Users/mpapadakis/legate.core/install/lib/python3.8/site-packages/legion_top.py", line 255, in run_path
    exec(code, module.__dict__, module.__dict__)
  File "a.py", line 13, in <module>
    print(x.reshape(1, 5).shape)
  File "/Users/mpapadakis/cunumeric/cunumeric/coverage.py", line 107, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/array.py", line 3030, in reshape
    thunk=self._thunk.reshape(computed_shape, order),
  File "/Users/mpapadakis/cunumeric/cunumeric/deferred.py", line 961, in reshape
    out_prod *= out_shape[out_dim]
IndexError: tuple index out of range
```

File: cunumeric/array.py (hunk 7 above)
Comment (manopapad) [reply cunumeric/array.py:3004]: Asking specifically about the `operator.index(.)` call

Comment (magnatelee) [reply cunumeric/array.py:3004]: that code window wasn't showing the line you were referring to. someone might try to do something like `arr.reshape(1, (2,3))`

**Review by manopapad — CHANGES_REQUESTED** [review]
One remaining check to add, otherwise looks good

File: cunumeric/array.py (hunk 3 above)
Comment (magnatelee) [reply cunumeric/array.py:3018]: [f7475cd](https://github.com/nv-legate/cunumeric/pull/490/commits/f7475cd3365cec6a3edf28ae094103497a4993ea)

**Comment by magnatelee** [comment]
@manopapad alright. I think I got a much better check now, so please take a look.

**Comment by magnatelee** [comment]
> I wrote this code to check all cases, there are two cases that throw ZeroDivisionError instead of ValueError:

should be fixed by 2d9982c

**Review by manopapad — APPROVED** [review]
Ship it

**Pull request merged**